### PR TITLE
Integrate ml-agent

### PR DESCRIPTION
**Pull request by jonoco**
- Created the PlayerAgent, which replaces the PlayerController for both user and AI input
- Changes to PlayerSupervisor are backwards compatible with PlayerController, i.e. PlayerController can still be used in place of PlayerAgent
- Added a training mode which quickly resets the scene to simplify training; it's toggled on the GameManager
- Added a rebound check, which looks for the ball becoming locked bouncing between boundaries, then changes its velocity but conserves its magnitude
- Added small randomization to ball launch direction
- Fixed the ball launch bug

### CODE REVIEW

**Comment by frich98** [comment]
I'm getting this error after I've setup this code for training. Looks like the ball Sounds definition has disappeared under the Ball objects. Can you look into this?
array_bounds_error

**Comment by jonoco** [comment]
It looks like the ball was expecting sounds to be set on the instance, I put in a check,
```
if (bounceSounds.Length > 0)
{
    AudioManager.Instance.PlaySound(bounceSounds[UnityEngine.Random.Range(0, bounceSounds.Length)]);
}
```

**Comment by frich98** [comment]
Looks like the Game Manager script now isn't happy. I made sure I'm pulling the code with your latest change above in it. 
image

**Comment by frich98** [comment]
Thank you for fixing those things! It's working for me now. Very cool to see it playing.

**Comment by frich98** [comment]
Something that might need to be addressed at some point but that doesn't seem to impact training - the score in the top left is cumulative and doesn't appear to reset after each game. I'm up to 15K now :)

File: Assets/Prefabs/Ball.prefab [hunk 1]
```diff
@@ -98,12 +98,13 @@ MonoBehaviour:
   m_EditorClassIdentifier: 
   paddle: {fileID: 0}
   hasStarted: 0
-  launchVelocity: {x: 2, y: 15}
-  bounceSounds:
-  - {fileID: 8300000, guid: 5cd521ee46ff3f54ea32bf02aef0fec0, type: 3}
-  - {fileID: 8300000, guid: c22a79a9053d9f849960377adb8b0f37, type: 3}
-  - {fileID: 8300000, guid: b5c32c79340bf5642b0e454004d80811, type: 3}
```
Comment (matzingc) [inline Assets/Prefabs/Ball.prefab]: Is there a reason for removing bounce sounds from the Ball prefab? I think they should remain there.

**Review by matzingc — CHANGES_REQUESTED** [review]

File: Assets/Prefabs/Ball.prefab (hunk 1 above)
Comment (jonoco) [reply Assets/Prefabs/Ball.prefab]: I fixed the sound references, I think they got killed accidentally in the merge. Prefab changes aren't as easy to follow as changes to the code.

**Review by matzingc — APPROVED** [review]

**Review by matzingc — COMMENTED** [review]
Thanks for making that requested change!

**Pull request merged**